include/openthread/mdns.h

```diff
@@ -388,6 +406,88 @@ otError otMdnsRegisterKey(otInstance            *aInstance,
  */
 otError otMdnsUnregisterKey(otInstance *aInstance, const otMdnsKey *aKey);
 
+/**
+ * Allocates a new iterator.
+ *
+ * An allocated iterator must be freed by the caller using `otMdnsFreeIterator()`.
+ *
+ * @param[in] aInstance    The OpenThread instance.
+ *
+ * @returns A pointer to the allocated iterator, or `NULL` if fails to allocate.
+ *
+ */
+otMdnsIterator *otMdnsAllocateIterator(otInstance *aInstance);
+
+/**
+ * Frees a previously allocated iterator.
+ *
+ * @param[in] aInstance    The OpenThread instance.
+ * @param[in] aIterator    The iterator to free.
+ *
+ */
+void otMdnsFreeIterator(otInstance *aInstance, otMdnsIterator *aIterator);
+
+/**
+ * Iterates over registered host entries.
+ *
+ * On success, @p aHost is populated with the next host information. Pointers within the `otMdnsHost` structure
+ * (like `mName`) remain valid until the next call to any OpenThread stack public or platform API/callback.
+ *
+ * @param[in]  aInstance   The OpenThread instance.
+ * @param[in]  aIterator   Pointer to the iterator to use.
+ * @param[out] aHost       Pointer to an `otMdnsHost` to return the information about next host entry.
+ * @param[out] aState      Pointer to an `otMdnsEntryState` to return the entry state.
+ *
+ * @retval OT_ERROR_NONE         @p aHost, @p aState, & @p aIterator are updated successfully.
+ * @retval OT_ERROR_NOT_FOUND    Reached the end of the list.
+ * @retval OT_ERROR_INVALID_ARG  Iterator is not valid.
+ *
+ */
+otError otMdnsGetNextHost(otInstance       *aInstance,
```

Comment:
Why do we need an iterator? otMndsGetNextHost should only get an otMdnsHost pointer and it should populate it with the corresponding values.
```otError otMdnsGetNextHost(otInstance *aInstance, otMdnsHost *aHost)```

Reply:
These functions are used to iterate over all the registered host/service/key entries. There can be more than one entry in the list. The iterator is used to keep track of where we are in the list.

Reply:
If you look at the SRP implementation you will see that it GetNext() type of functions require a pointer only. If the pointer is null it will provide the first item in the list if not the next item (since every item contains a mNext pointer). So the entry itself is the iterator. But removing the extra class we keep the code cleaner and smaller.

Reply:
This is not applicable. We do not want to force the internal data structure used to store information in the mDNS module to match the parameter format of API calls (`otMdnsHost/Service/Key` structures). This approach provides flexibility in updating the internal data model.

The iterator is also necessary to track any heap-allocated buffers used when populating the `otMdnsService`. This ensures that these allocations are properly freed.

The SRP client intentionally delegates all structure/buffer management to the API caller. This design decision aims to increase the SRP client implementation's flexibility at the cost of increased complexity for the API user. This approach was selected since the SRP client can be used on devices with very limited resources.